### ISSUE TITLE
Is FrameEval threaded or parallel processing in any way?

### ISSUE THREAD

**Issue by rlaphoenix** [vapoursynth/vapoursynth#890]
I'm trying to effectively run a function across a clip, using two GPU devices.

For example:
```py
import vapoursynth as vs
from vsgan import ESRGAN
core = vs.core

c = core.lsmas.LWLibavSource(source="./input.mkv")

c = c.fmtc.resample (css="444")
c = c.fmtc.matrix (mat="601", col_fam=vs.RGB)
c = c.fmtc.bitdepth (bits=16)

ca = core.std.SelectEvery(clip=c, cycle=2, offsets=0)
cb = core.std.SelectEvery(clip=c, cycle=2, offsets=1)

ca = ESRGAN(ca, device="cuda:0").load(r'./animevideov3.pth').apply().clip
cb = ESRGAN(cb, device="cuda:1").load(r'./realesr-animevideov3.pth').apply().clip

c = core.std.Interleave(clips=[ca, cb]);

c = c.fmtc.matrix (mat="601", col_fam=vs.YUV, bits=16)
c = c.fmtc.resample (css="420")
c = c.fmtc.bitdepth (bits=10)

c.set_output()
```

This example works and performance is as expected.

However, within `ESRGAN` I am able to change that `device` parameter internally when needed within the FrameEval call based on n (via n % 2 as a list index to choose a device).

You can see here what it's doing: https://github.com/rlaphoenix/VSGAN/blob/04ffe6ef1b3a1c8bb84dd29fedb7bab5a0b24bf5/vsgan/networks/esrgan.py#L105

tldr;

```py
FrameEval(clip, clip, frame_eval_body)

def frame_eval_body(n, ...):
    i = n % 2
    device = devices[i]
    model = model[i]
    # ... uses model and device
```

This works, and changes as expected, but performance is nowhere near expected. It's about half what the above example gets.
This is what my GPU load looks like with the above example:

![image](https://user-images.githubusercontent.com/17136956/179827868-b60d85ae-9ef5-410c-93ac-0a74a1acbd9c.png)

And here is what it looks like when just choosing based on n%2 within the FrameEval:

![image](https://user-images.githubusercontent.com/17136956/179828847-8f1859c4-7781-411a-877e-864f4a1f75e6.png)
![image](https://user-images.githubusercontent.com/17136956/179828956-dc2c10bb-af49-4029-b7ef-a0c0b3af8518.png)

I imagine the performance is shallow because it's switching between devices, and frames arent executed in parallel.
Is there a way to safely execute at least 2 frames in parallel sufficiently for this goal of multi-GPU processing?

**Comment by myrsloik** [comment]
FrameEval isn't fully parallel since it runs Python code so it'll never do what you want. But why can't you have 2 filter instances?

**Comment by rlaphoenix** [comment]
So there's no way to multi-thread it at all? Because in the method that actually utilizes the GPUs, I only get that load if I set core.num_threads to an absurd number (like 50 for good load to last until 150~ frames, or 100 to last until 370~ frames).

Ideally, I'd like a way for FrameEval to be multi-threaded in some way rather than using FrameEval twice as shown earlier, as it would overall reduce the amount of code needing to be done on my end, and would be cleaner.

**Comment by cid-chan** [comment]
FrameEval runs Python Code. Python has the infamous GIL that lets the interpreter only run one thing at the same time.* If you want to parallelize your code, get rid of as many ModifyFrame and FrameEval-calls as possible

*IO and sometimes FFI exempted.

**Comment by YomikoR** [comment]
Are we unable to write a C program that calls FrameEval?

**Comment by myrsloik** [comment]
> Are we unable to write a C program that calls FrameEval?

Sure, if you skip python and use the C api directly it'll work
